File: archivist/src/main/java/com/zorroa/archivist/web/api/AssetController.java
package com.zorroa.archivist.web.api;

import com.google.common.collect.ImmutableMap;
import com.zorroa.archivist.HttpUtils;
import com.zorroa.archivist.domain.*;
import com.zorroa.archivist.security.SecurityUtils;
import com.zorroa.archivist.service.*;
import com.zorroa.archivist.web.MultipartFileSender;
import com.zorroa.common.elastic.ElasticClientUtils;
import com.zorroa.sdk.client.exception.ArchivistWriteException;
import com.zorroa.sdk.domain.*;
import com.zorroa.sdk.filesystem.ObjectFileSystem;
import com.zorroa.sdk.schema.ProxySchema;
import com.zorroa.sdk.search.AssetSearch;
import com.zorroa.sdk.search.AssetSuggestBuilder;
import org.apache.tika.Tika;
import org.elasticsearch.ResourceNotFoundException;
import org.elasticsearch.action.get.GetResponse;
import org.elasticsearch.action.search.SearchResponse;
import org.elasticsearch.action.suggest.SuggestResponse;
import org.elasticsearch.client.Client;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import javax.servlet.ServletOutputStream;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.validation.Valid;
import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;
import java.util.Set;

@RestController
public class AssetController {

    private static final Logger logger = LoggerFactory.getLogger(AssetController.class);

    @Value("${zorroa.cluster.index.alias}")
    private String alias;

    @Autowired
    Client client;

    @Autowired
    AssetService assetService;

    @Autowired
    NoteService noteService;

    @Autowired
    SearchService searchService;

    @Autowired
    EventLogService logService;

    @Autowired
    ImageService imageService;

    @Autowired
    ObjectFileSystem ofs;

    @Autowired
    CommandService commandService;

    private static final Tika tika = new Tika();

    /**
     * Describes a file to stream.
     */
    private static class StreamFile {
        public String path;
        public String mimeType;
        public boolean proxy;

        public StreamFile(String path, String mimeType, boolean proxy) {
            this.path = path;
            this.mimeType = mimeType;
            this.proxy = proxy;
        }
    }

    /**
     * We could try to detect it using something like Tika but
     * there are only a couple types.
     */
    private static final Map<String,String> PROXY_MIME_LOOKUP =
            ImmutableMap.of("png", "image/png",
                            "jpg", "image/jpeg");

    public StreamFile getPreferredFormat(Asset asset, String preferExt, boolean fallback, boolean streamProxy) {
        if (streamProxy) {
            return getProxyStream(asset);
        }
        else {
            String path = asset.getAttr("source.path", String.class);
            String mediaType = asset.getAttr("source.mediaType", String.class);

            if (preferExt != null) {
                path = path.substring(0, path.lastIndexOf('.')+1) + preferExt;
                mediaType = tika.detect(path);
            }

            if (new File(path).exists()) {
                return new StreamFile(path, mediaType,false);
            } else {
                if (fallback) {
                    return getProxyStream(asset);
                }
                else {
                    return null;
                }
            }
        }
    }

    public StreamFile getProxyStream(Asset asset) {
        Proxy largestProxy = asset.getProxies().getLargest();
        return new StreamFile(
                ofs.get(largestProxy.getId()).getFile().toString(),
                PROXY_MIME_LOOKUP.getOrDefault(largestProxy.getFormat(),
                        "application/octet-stream"), true);
    }

    @RequestMapping(value = "/api/v1/assets/{id}/_stream", method = RequestMethod.GET)
    public void streamAsset(@RequestParam(defaultValue="true", required=false) Boolean fallback, @RequestParam(value="ext", required=false) String ext, @PathVariable String id, HttpServletRequest request, HttpServletResponse response) throws Exception {

        Asset asset = assetService.get(id);
        boolean canExport = SecurityUtils.canExport(asset);
        StreamFile format = getPreferredFormat(asset, ext, fallback, !canExport);

        /*
         * Nothing to return...
         */
        if (format == null) {
            response.setStatus(404);
        }
        else {
            try {
                MultipartFileSender.fromPath(Paths.get(format.path))
                        .with(request)
                        .with(response)
                        .setContentType(format.mimeType)
                        .serveResource();
                if (canExport) {
                    logService.logAsync(UserLogSpec.build(LogAction.View, "asset", asset.getId()));
                }
            } catch (Exception e) {
                logger.warn("MultipartFileSender failed on {}, unexpected {}", id, e.getMessage());
            }
        }
    }

    @RequestMapping(value="/api/v1/assets/{id}/notes", method=RequestMethod.GET)
    public List<Note> getNotes(@PathVariable String id) throws IOException {
        return noteService.getAll(id);
    }

    @RequestMapping(value="/api/v1/assets/{id}/proxies/closest/{size:\\d+x\\d+}", method=RequestMethod.GET)
    public ResponseEntity<InputStreamResource> getClosestProxy(@PathVariable String id, @PathVariable(required=false) String size) throws IOException {
        try {
            String[] wh = size.split("x");
            ProxySchema proxies = assetService.get(id).getProxies();
            Proxy proxy = proxies.getClosest(Integer.valueOf(wh[0]), Integer.valueOf(wh[1]));
            if (proxy == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
            }
            return imageService.serveImage(proxy);
        } catch (Exception e) {
            throw new ResourceNotFoundException(e.getMessage());
        }
    }

    @RequestMapping(value="/api/v1/assets/{id}/proxies/largest", method=RequestMethod.GET)
    public ResponseEntity<InputStreamResource> getLargestProxy(@PathVariable String id) throws IOException {
        try {
            ProxySchema proxies = assetService.get(id).getProxies();
            Proxy proxy = proxies.getLargest();
            if (proxy == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
            }
            return imageService.serveImage(proxy);
        } catch (Exception e) {
            throw new ResourceNotFoundException(e.getMessage());
        }
    }

    @RequestMapping(value="/api/v1/assets/{id}/proxies/smallest", method=RequestMethod.GET)
    public ResponseEntity<InputStreamResource> getSmallestProxy(@PathVariable String id) throws IOException {
        try {
            ProxySchema proxies = assetService.get(id).getProxies();
            Proxy proxy = proxies.getSmallest();
            if (proxy == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
            }
            return imageService.serveImage(proxy);
        } catch (Exception e) {
            throw new ResourceNotFoundException(e.getMessage());
        }
    }

    @RequestMapping(value="/api/v2/assets/_search", method=RequestMethod.POST)
    public void searchV2(@RequestBody AssetSearch search, HttpServletResponse httpResponse) throws IOException {
        httpResponse.setContentType(MediaType.APPLICATION_JSON_VALUE);
        SearchResponse response = searchService.search(search);
        HttpUtils.writeElasticResponse(response, httpResponse);
    }

    @RequestMapping(value="/api/v3/assets/_search", method=RequestMethod.POST)
    public PagedList<Asset> searchV3(@RequestBody AssetSearch search) throws IOException {
        return searchService.search(new Pager(search.getFrom(), search.getSize(), 0), search);
    }

    @RequestMapping(value="/api/v4/assets/_search", method=RequestMethod.POST)
    public void searchV4(@RequestBody AssetSearch search, ServletOutputStream out) throws IOException {
        searchService.search(new Pager(search.getFrom(), search.getSize(), 0), search, out);
    }

    @RequestMapping(value="/api/v1/assets/_fields", method=RequestMethod.GET)
    public Map<String, Set<String>> getFields() throws IOException {
        return searchService.getFields();
    }

    @RequestMapping(value="/api/v1/assets/_fields/hide", method=RequestMethod.PUT)
    public Object unhideField(@RequestBody HideField update) throws IOException {
        return HttpUtils.status("field", "hide",
                searchService.updateField(update.setHide(true).setManual(true)));
    }

    @RequestMapping(value="/api/v1/assets/_fields/hide", method=RequestMethod.DELETE)
    public Object hideField(@RequestBody HideField update) throws IOException {
        return HttpUtils.status("field", "unhide",
                searchService.updateField(update.setHide(false)));
    }

    @RequestMapping(value="/api/v1/assets/_mapping", method=RequestMethod.GET)
    public Map<String,Object> getMapping() throws IOException {
        return assetService.getMapping();
    }

    @RequestMapping(value="/api/v2/assets/_count", method=RequestMethod.POST, produces=MediaType.APPLICATION_JSON_VALUE)
    public Object count(@RequestBody AssetSearch search) throws IOException {
        return HttpUtils.count(searchService.count(search));
    }

    @RequestMapping(value="/api/v1/assets/{id}/_exists", method=RequestMethod.GET)
    public Object exists(@PathVariable String id) throws IOException {
        return HttpUtils.exists(id, assetService.exists(id));
    }

    @RequestMapping(value="/api/v2/assets/_suggest", method=RequestMethod.POST, produces=MediaType.APPLICATION_JSON_VALUE)
    public String suggestV2(@RequestBody AssetSuggestBuilder builder) throws IOException {
        SuggestResponse response = searchService.suggest(builder.getText());
        return response.toString();
    }

    @RequestMapping(value="/api/v3/assets/_suggest", method=RequestMethod.POST)
    public Object suggestV3(@RequestBody AssetSuggestBuilder suggest) throws IOException {
        return searchService.getSuggestTerms(suggest.getText());
    }

    @RequestMapping(value="/api/v1/assets/{id}", method=RequestMethod.GET)
    public void get(@PathVariable String id, HttpServletResponse httpResponse) throws IOException {
        GetResponse response = client.prepareGet(alias, "asset", id).get();
        HttpUtils.writeElasticResponse(response, httpResponse);
    }

    @RequestMapping(value="/api/v2/assets/{id}", method=RequestMethod.GET)
    public Object getV2(@PathVariable String id) {
        return assetService.get(id);
    }

    @RequestMapping(value="/api/v1/assets/_path", method=RequestMethod.GET)
    public Object getByPath(@RequestBody Map<String,String> path) {
        return assetService.get(path.get("path"));
    }

    @RequestMapping(value="/api/v1/assets/{id}/_elements", method=RequestMethod.GET)
    public PagedList<Document> getElements(@PathVariable String id,
                                           @RequestParam(value="from", required=false) Integer from,
                                           @RequestParam(value="count", required=false) Integer count) {
        return assetService.getElements(id, new Pager(from, count));
    }

    @RequestMapping(value="/api/v1/assets/{id}", method=RequestMethod.DELETE)
    public Object delete(@PathVariable String id) throws IOException {
        Asset asset = assetService.get(id);
        if (!SecurityUtils.hasPermission("write", asset)) {
            throw new ArchivistWriteException("delete access denied");
        }

        boolean result = assetService.delete(id);
        return HttpUtils.deleted("asset", id, result);
    }

    @RequestMapping(value="/api/v1/assets/{id}", method=RequestMethod.PUT, produces=MediaType.APPLICATION_JSON_VALUE)
    public Object update(@RequestBody Map<String, Object> attrs, @PathVariable String id) throws IOException {
        Asset asset = assetService.get(id);
        if (!SecurityUtils.hasPermission("write", asset)) {
            throw new ArchivistWriteException("update access denied");
        }

        assetService.update(id, attrs);
        return HttpUtils.updated("asset", id, true, assetService.get(id));
    }

    @RequestMapping(value="/api/v1/assets/{id}/_fields", method=RequestMethod.DELETE)
    public Object removeFields(@RequestBody Set<String> fields, @PathVariable String id) throws IOException {
        assetService.removeFields(id, fields);
        return HttpUtils.updated("asset", id, true, assetService.get(id));
    }

    @RequestMapping(value="/api/v1/assets/_index", method=RequestMethod.POST, produces=MediaType.APPLICATION_JSON_VALUE)
    public AssetIndexResult index(@RequestBody AssetIndexSpec spec) throws IOException {
        return assetService.index(spec);
    }

    public static class SetPermissionsRequest {
        public AssetSearch search;
        public Acl acl;

    }

    @PreAuthorize("hasAuthority('group::share') || hasAuthority('group::administrator')")
    @RequestMapping(value="/api/v1/assets/_permissions", method=RequestMethod.PUT)
    public Command setPermissions(
            @Valid @RequestBody SetPermissionsRequest req) throws Exception {
        CommandSpec spec = new CommandSpec();
        spec.setType(CommandType.UpdateAssetPermissions);
        spec.setArgs(new Object[] {
                req.search,
                req.acl
        });
        return commandService.submit(spec);
    }

    @RequestMapping(value="/api/v1/refresh", method=RequestMethod.PUT)
    public void refresh() {
        ElasticClientUtils.refreshIndex(client, 0);
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/web/api/AssetController.kt
package com.zorroa.archivist.web.api

import com.google.common.cache.CacheBuilder
import com.google.common.cache.CacheLoader
import com.zorroa.archivist.util.HttpUtils
import com.zorroa.archivist.util.HttpUtils.CACHE_CONTROL
import com.zorroa.archivist.domain.*
import com.zorroa.archivist.repository.AssetIndexResult
import com.zorroa.archivist.search.AssetSearch
import com.zorroa.archivist.search.AssetSuggestBuilder
import com.zorroa.archivist.security.canExport
import com.zorroa.archivist.security.hasPermission
import com.zorroa.archivist.service.*
import com.zorroa.archivist.util.event
import com.zorroa.archivist.web.MultipartFileSender
import com.zorroa.archivist.web.sender.FlipbookSender
import com.zorroa.common.domain.ArchivistWriteException
import com.zorroa.common.schema.Proxy
import com.zorroa.common.schema.ProxySchema
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.InputStreamResource
import org.springframework.http.CacheControl
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.io.IOException
import java.net.URI
import java.nio.file.Paths
import java.util.*
import java.util.concurrent.TimeUnit
import javax.servlet.ServletOutputStream
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@RestController
class AssetController @Autowired constructor(
        private val indexService: IndexService,
        private val assetService: AssetService,
        private val searchService: SearchService,
        private val folderService: FolderService,
        private val imageService: ImageService,
        private val fieldService: FieldService,
        private val fileServerProvider: FileServerProvider,
        private val fileStorageService: FileStorageService
){

    @GetMapping(value = ["/api/v1/assets/_fields"])
    fun getFields(response: HttpServletResponse) : Map<String, Set<String>> {
        response.setHeader("Cache-Control", CacheControl.maxAge(
                30, TimeUnit.SECONDS).cachePrivate().headerValue)
        return fieldService.getFields("asset")
    }

    val mapping: Map<String, Any>
        @GetMapping(value = ["/api/v1/assets/_mapping"])
        @Throws(IOException::class)
        get() = indexService.getMapping()


    fun getPreferredFormat(asset: Document, forceProxy: Boolean): ServableFile? {
        if (forceProxy) {
            val proxy = getProxyStream(asset)
             if (proxy != null) {
                 return fileServerProvider.getServableFile(URI(proxy.uri))
             }

        } else  {
            return fileServerProvider.getServableFile(fileServerProvider.getStorageUri(asset))
        }

        return null
    }

    fun getProxyStream(asset: Document): FileStorage? {
        // If the file doesn't have a proxy this will throw.
        val proxies = asset.getAttr("proxies", ProxySchema::class.java)

        if (proxies != null) {
            val largest = proxies.getLargest()
            if (largest != null) {
                fileStorageService.get(largest.id!!)
            }
        }
        return null
    }

    @RequestMapping(value = ["/api/v1/assets/{id}/_stream"], method = [RequestMethod.GET, RequestMethod.HEAD])
    @Throws(Exception::class)
    fun streamAsset(@RequestParam(defaultValue = "true", required = false) fallback: Boolean,
                    @RequestParam(value = "ext", required = false) ext: String?,
                    @PathVariable id: String, request: HttpServletRequest, response: HttpServletResponse) {

        val asset = indexService.get(id)
        val canExport = canExport(asset)
        val ofile = getPreferredFormat(asset, !canExport)

        if (ofile == null) {
            response.status = 404
        }
        else {
            if (request.method == "HEAD") {
                /**
                 * Only non-local files need to be signed.
                 */
                if (!ofile.isLocal()) {
                    response.setHeader("X-Zorroa-Signed-URL", ofile.getSignedUrl().toString())
                }
            }
            else {
                try {
                    logger.event("view Asset", mapOf("assetId" to asset.id))
                    if (!ofile.isLocal()) {
                        ofile.copyTo(response)
                    } else {
                        MultipartFileSender.fromPath(ofile.getLocalFile())
                                .with(request)
                                .with(response)
                                .setContentType(ofile.getStat().mediaType)
                                .serveResource()

                    }
                } catch (e: Exception) {
                    response.sendError(404, "StorageSystem unable to find file")
                }
            }
        }
    }

    private val proxyLookupCache = CacheBuilder.newBuilder()
            .maximumSize(10000)
            .expireAfterWrite(1, TimeUnit.HOURS)
            .build(object : CacheLoader<String, Proxy>() {
                @Throws(Exception::class)
                override fun load(slug: String): Proxy? {
                    val e = slug.split(":")
                    val proxies = indexService.getProxies(e[1])

                    return when {
                        e[0] == "closest" -> proxies.getClosest(e[2].toInt(), e[3].toInt())
                        e[0] == "atLeast" -> proxies.atLeastThisSize(e[2].toInt())
                        e[0] == "smallest" -> proxies.getSmallest()
                        e[0] == "largest" -> proxies.getLargest()
                        else -> proxies.getLargest()
                    }
                }
            })

    @GetMapping(value = ["/api/v1/assets/{id}/proxies/closest/{width:\\d+}x{height:\\d+}"])
    @Throws(IOException::class)
    fun getClosestProxy(req: HttpServletRequest,
                        rsp: HttpServletResponse,
                        @PathVariable id: String,
                        @PathVariable width: Int,
                        @PathVariable height: Int) {
        return try {
            imageService.serveImage(req, rsp, proxyLookupCache.get("closest:$id:$width:$height"))
        } catch (e: Exception) {
            rsp.status = HttpStatus.NOT_FOUND.value()
        }
    }

    @GetMapping(value = ["/api/v1/assets/{id}/proxies/atLeast/{size:\\d+}"])
    @Throws(IOException::class)
    fun getAtLeast(req: HttpServletRequest,
                   rsp: HttpServletResponse,
                   @PathVariable id: String,
                   @PathVariable(required = true) size: Int) {
        try {
            imageService.serveImage(req, rsp, proxyLookupCache.get("atLeast:$id:$size"))
        } catch (e: Exception) {
            rsp.status = HttpStatus.NOT_FOUND.value()
        }
    }

    @GetMapping(value = ["/api/v1/assets/{id}/proxies/largest"])
    @Throws(IOException::class)
    fun getLargestProxy(req: HttpServletRequest,
                        rsp: HttpServletResponse,
                        @PathVariable id: String) {
        try {
            imageService.serveImage(req, rsp, proxyLookupCache.get("largest:$id"))
        } catch (e: Exception) {
            rsp.status = HttpStatus.NOT_FOUND.value()
        }
    }

    @GetMapping(value = ["/api/v1/assets/{id}/proxies/smallest"])
    @Throws(IOException::class)
    fun getSmallestProxy(req: HttpServletRequest,
                         rsp: HttpServletResponse,
                         @PathVariable id: String) {
        return try {
            imageService.serveImage(req, rsp, proxyLookupCache.get("smallest:$id"))
        } catch (e: Exception) {
            rsp.status = HttpStatus.NOT_FOUND.value()
        }
    }

    @PostMapping(value = ["/api/v3/assets/_search"])
    @Throws(IOException::class)
    fun searchV3(@RequestBody search: AssetSearch): PagedList<Document> {
        return searchService.search(Pager(search.from, search.size, 0), search)
    }

    @PostMapping(value = ["/api/v4/assets/_search"])
    @Throws(IOException::class)
    fun searchV4(@RequestBody search: AssetSearch, out: ServletOutputStream) {
        searchService.search(Pager(search.from, search.size, 0), search, out)
    }

    @PutMapping(value = ["/api/v1/assets/_fields/hide"])
    @Throws(IOException::class)
    fun unhideField(@RequestBody update: HideField): Any {
        return HttpUtils.status("field", "hide",
                fieldService.updateField(update.setHide(true).setManual(true)))
    }

    @DeleteMapping(value = ["/api/v1/assets/_fields/hide"])
    @Throws(IOException::class)
    fun hideField(@RequestBody update: HideField): Any {
        return HttpUtils.status("field", "unhide",
                fieldService.updateField(update.setHide(false)))
    }

    @PostMapping(value = ["/api/v2/assets/_count"])
    @Throws(IOException::class)
    fun count(@RequestBody search: AssetSearch): Any {
        return HttpUtils.count(searchService.count(search))
    }

    @GetMapping(value = ["/api/v1/assets/{id}/_exists"])
    @Throws(IOException::class)
    fun exists(@PathVariable id: String): Any {
        return HttpUtils.exists(id, indexService.exists(id))
    }

    @PostMapping(value = ["/api/v3/assets/_suggest"])
    @Throws(IOException::class)
    fun suggestV3(@RequestBody suggest: AssetSuggestBuilder): Any {
        return searchService.getSuggestTerms(suggest.text)
    }

    @GetMapping(value = ["/api/v2/assets/{id}"])
    fun getV2(@PathVariable id: String): Any {
        return indexService.get(id)
    }

    @GetMapping(value = ["/api/v1/assets/_path"])
    fun getByPath(@RequestBody path: Map<String, String>): Document? {
        return path["path"]?.let { indexService.get(Paths.get(it)) }
    }

    @DeleteMapping(value = ["/api/v1/assets/{id}"])
    @Throws(IOException::class)
    fun delete(@PathVariable id: String): Any {
        return HttpUtils.deleted("asset", id, assetService.delete(id))
    }

    @DeleteMapping(value = ["/api/v1/assets"])
    @Throws(IOException::class)
    fun batchDelete(@RequestBody batch: BatchDeleteAssetsRequest): BatchDeleteAssetsResponse {
        return assetService.batchDelete(batch.assetIds)
    }

    @PutMapping(value = ["/api/v1/assets/{id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    @Throws(IOException::class)
    fun update(@RequestBody attrs: Map<String, Any>, @PathVariable id: String): Any {
        val asset = indexService.get(id)
        if (!hasPermission("write", asset)) {
            throw ArchivistWriteException("update access denied")
        }

        indexService.update(id, attrs)
        return HttpUtils.updated("asset", id, true, indexService.get(id))
    }

    @DeleteMapping(value = ["/api/v1/assets/{id}/_fields"])
    @Throws(IOException::class)
    fun removeFields(@RequestBody fields: MutableSet<String>, @PathVariable id: String): Any {
        indexService.removeFields(id, fields)
        return HttpUtils.updated("asset", id, true, indexService.get(id))
    }

    @GetMapping(value = ["/api/v1/assets/{id}/_clipChildren"])
    @Throws(IOException::class)
    fun clipChildren(@PathVariable id: String, rsp: HttpServletResponse) {
        FlipbookSender(id, searchService).serveResource(rsp)
    }

    @PostMapping(value = ["/api/v1/assets/_index"], produces = [(MediaType.APPLICATION_JSON_VALUE)])
    @Throws(IOException::class)
    fun index(@RequestBody spec: AssetIndexSpec): AssetIndexResult {
        return indexService.index(spec)
    }

    class SetFoldersRequest {
        var folders: List<UUID>? = null
    }

    /**
     * Reset all folders for a given asset.  Currently only used for syncing.
     */
    @PreAuthorize("hasAuthority(T(com.zorroa.security.Groups).ADMIN)")
    @PutMapping(value = ["/api/v1/assets/{id}/_setFolders"])
    @Throws(Exception::class)
    fun setFolders(@PathVariable id: String, @RequestBody req: SetFoldersRequest): Any {
        req?.folders?.let {
            folderService.setFoldersForAsset(id, it)
            return HttpUtils.updated("asset", id, true)
        }
        return HttpUtils.updated("asset", id, false)
    }

    class SetPermissionsRequest {
        var search: AssetSearch? = null
        var acl: Acl? = null

    }

    /*
    @PreAuthorize("hasAuthority(T(com.zorroa.security.Groups).SHARE) || hasAuthority(T(com.zorroa.security.Groups).ADMIN)")
    @PutMapping(value = ["/api/v1/assets/_permissions"])
    @Throws(Exception::class)
    fun setPermissions(
            @Valid @RequestBody req: SetPermissionsRequest): Command {
        val spec = CommandSpec()
        spec.type = CommandType.UpdateAssetPermissions
        spec.args = arrayOf(req.search, req.acl)
        return commandService.submit(spec)
    }
    */

    @PutMapping(value = ["/api/v1/refresh"])
    fun refresh() {
        logger.warn("Refresh called.")
    }
    companion object {

        private val logger = LoggerFactory.getLogger(AssetController::class.java)
    }
}
